go/vt/wrangler/reparent.go

```diff
@@ -996,55 +995,279 @@ func (wr *Wrangler) emergencyReparentShardLocked(ctx context.Context, ev *events
+			if err != nil {
+				return nil, err
+			}
+			positionMap[alias] = executedFilePosition
+		}
+	}
+
+	if !*gtidBased {
+		// If all positions are not comparable, we can't do file based comparisons at all.
+		filePositionList := make([]mysql.Position, 0, len(replicationStatusMap))
+		for _, p := range positionMap {
+			filePositionList = append(filePositionList, p)
+		}
+		if !mysql.AllPositionsComparable(filePositionList) {
+			return nil, fmt.Errorf("we can't compare all file based positions")
+		}
+	}
+
+	// Find a position that is fully caught up when errant GTIDs are taken out of the equation.
+	var winningPosition mysql.Position
+	for _, position := range positionMap {
+		if winningPosition.IsZero() {
+			winningPosition = position
+		}
+		if position.AtLeast(winningPosition) {
+			winningPosition = position
+		}
+	}
+
+	// Now that we have a high water mark, we can get a list of all tablets that are errant GTID free
+	// that hit this high water mark.
+	var validCandidates []string
+	for alias, position := range positionMap {
+		if badTablets.Has(alias) {
+			// We should exclude tablets we've already determined have errant GTIDs from consideration.
+			continue
+		}
+
+		if position.AtLeast(winningPosition) {
+			validCandidates = append(validCandidates, alias)
+		}
+	}
+
+	return validCandidates, nil
+}
+
+func (wr *Wrangler) stopReplicationAndBuildStatusMaps(ctx context.Context, ev *events.Reparent, tabletMap map[string]*topo.TabletInfo, waitReplicasTimeout time.Duration, unreachableReplicas sets.String) (map[string]*replicationdatapb.StopReplicationStatus, map[string]*replicationdatapb.MasterStatus, error) {
+	// Stop replication on all replicas, get their current
+	// replication position
+	event.DispatchUpdate(ev, "stop replication on all replicas")
+	statusMap := make(map[string]*replicationdatapb.StopReplicationStatus)
+	masterStatusMap := make(map[string]*replicationdatapb.MasterStatus)
+	mu := sync.Mutex{}
+
+	errChan := make(chan error)
+	groupCtx, groupCancel := context.WithTimeout(ctx, waitReplicasTimeout)
+	defer groupCancel()
+	fillStatus := func(alias string, tabletInfo *topo.TabletInfo) {
+		err := fmt.Errorf("fillStatus did not successfully complete")
+		returnErr := func() { errChan <- err }
+		defer returnErr()
+
+		wr.logger.Infof("getting replication position from %v", alias)
+		ctx, cancel := context.WithCancel(groupCtx)
+		defer cancel()
+		var stopReplicationStatus *replicationdatapb.StopReplicationStatus
+		_, stopReplicationStatus, err = wr.tmc.StopReplicationAndGetStatus(ctx, tabletInfo.Tablet, replicationdatapb.StopReplicationMode_IOTHREADONLY)
+		switch err {
+		case mysql.ErrNotReplica:
+			fmt.Printf("Found ErrNotReplica for alias: %v", alias)
+			masterStatus, err := wr.tmc.DemoteMaster(ctx, tabletInfo.Tablet)
+			if err != nil {
+				wr.logger.Warningf("replica %v thinks it's master but we failed to demote it", alias)
+				return
```

Comment:
The `err` in this scope is shadowing the one that gets sent to `errChan`, so we won't send back the real error from DemoteMaster. In general, this pattern of deferring the channel send inside the function is hard to follow. What if we do something like this?

```go
fillStatus := func(...) error {
  ...
}

errChan := make(chan error)
for ... {
  go func() {
    errChan <- fillStatus(...)
  }()
}
```

Reply:
In this case, wouldn't there be no guarantee that `fillStatus` would return? I wrote the code with the defer error to ensure we definitely get back a response no matter what. I would have used named vars, but I don't think you can use them with closures in go? And without a closure, this requires quite a lot of argument passing. Thoughts?

Reply:
Now that you have explained it, this pattern is fine with the shadowing fixed.

Reply:
The shadowing is fixed now.